mod_s3_upload/src/mod_s3_upload.erl

```diff
@@ -463,15 +466,17 @@ decorated_put_url(UploadRequest, ServiceParams, URL) ->
           ObjectURL :: binary().
 % generate a unique random object URL for the given filename
 object_url(BucketURL, ObjectName) ->
-    #{path := BasePath} = UriMap = uri_string:parse(BucketURL),
-    uri_string:recompose(UriMap#{path => <<BasePath/binary, "/", ObjectName/binary>>}).
+    uri_string:resolve(ObjectName, BucketURL).
 
 -spec object_name(
         Filename :: binary()
        ) ->
           ObjectName :: binary().
 % generate reasonably unique sortable (by time first) object name.
 object_name(Filename) ->
-    str:format("~.36B~.36B-~s", [os:system_time(microsecond),
-                                 erlang:phash2(node()),
-                                 Filename]).
+    <<RandomPrefix:128/big-unsigned-integer>> = crypto:strong_rand_bytes(16),
+    RandomPrefixS = str:format("~32.16.0b", [RandomPrefix]),
+    FolderName = str:substr(RandomPrefixS, 1, 2),
+    str:format("~s/~s-~s", [FolderName,
```

Comment:
I've actually placed the `/` separator before the filename for my deployment (as clients such as Gajim and Dino display this random-looking hash to an end-user).

Reply:
@jpds While there isn't actually any true hierarchy in most S3CS implementations (e.g. there are no tangible folders, just the separators), generally all storage browsers will treat each prefix segment as a folder, which is why the original version of `mod_s3_upload` did not place a `/` after the disambiguator.

What's going on here is that a small slice of the disambiguator/slug is now being used as a prefix to simplify browsing through a large collection of files, e.g. `DE/DEADBEEF-cat.png`.

Reply:
The purpose here isn’t actually ease of browsing. Many S3 implementations, including AWS’ and Ceph’s, can experience performance issues when many requests hit the same prefix. This does some sharding of the files between prefixes by prepending that random byte from the distinguisher.

Reply:
@RomanHargrave I know, it's a tiny detail, however - I prefer not to change the filename the user has set as the prefix is displayed to them in certain apps, for instance Dino does this before it is downloaded and displayed:

![image](https://github.com/processone/ejabberd-contrib/assets/29158971/28cb6a5d-4319-478d-a07d-23cf35b18d47)